### PR TITLE
e2e(angular-aggiemap): Implemented user-flow oriented testing

### PR DESCRIPTION
**Pull request by Minhb171**
Corrected errors that were existing previously and completed main user-flows for the desktop and mobile version of the Aggie Map. Testing was done by stubbing server responses with predetermined data to avoid reliance on receiving actual data from the server.

Main user-flows have been determined to be:

- Icons appearing for each map layer when clicked
- Search bar results
- Map Rotation
- Checking if pop-ups are displayed correctly for varying clickable objects
- Routes being displayed on the map
- Zoom functionality
- Links bringing user to correct page

### CODE REVIEW

File: apps/aggiemap-angular-e2e/src/integration/features-layers.spec.ts [hunk 1]
```diff
@@ -18,75 +26,222 @@ desktopSizes.forEach((size) => {
...
+        cy.get('canvas').trigger('mouseover').click(900, 650)
+      }
+      else if (size[0] === 1366) {
+        cy.wait(2000)
+        cy.get('canvas').trigger('mouseover').click(625, 500)
+      }
+      else if (size[0] === 1440) {
+        cy.wait(2000)
+        cy.get('canvas').trigger('mouseover').click(670, 555)
+      }
```
Comment (Proantagonist) [inline apps/aggiemap-angular-e2e/src/integration/features-layers.spec.ts:43]: This is subject to breaking if the features on the screen coordinates are removed. For construction shapes, this happens quite often. I don't know a better way off the top of my head but try and look into accessing component state through cypress. If you can do that, you can maybe grab the map instance and query the screen and retrieve the screen coordinates of a shape that is rendered within the viewport.

File: apps/aggiemap-angular-e2e/src/integration/mobile-layers.spec.ts [hunk 2]
```diff
@@ -1,10 +1,372 @@
...
+      // check if route duration and length is displayed
+      cy.get('.directions-overview').should('be.visible')
+      // slide directions back down to access other route options
+      cy.get('tamu-gisc-trip-planner-directions-mobile').move({ x: 0, y: 600, position: 'center', force: true })
+
+      // test bike route option
+      cy.wait(2000)
+      cy.get('.travel-modes > :nth-child(2) > div')
+          .click()
+          .should('have.class', 'active')
```
Comment (Proantagonist) [inline apps/aggiemap-angular-e2e/src/integration/mobile-layers.spec.ts:109]: As mentioned here, routing testing is totally unnecessary here and especially for many map features

File: apps/aggiemap-angular-e2e/src/support/index.d.ts [hunk 3]
```diff
@@ -21,5 +21,9 @@ declare namespace Cypress {
      checkLink(elementText: string, link: string): Chainable<Element>
      containsAnyText(element: string): Chainable<Element>
      checkMenuItem(num: string, name:string): Chainable<Element>
+     checkPopUp(): Chainable<Element>
+     checkNavPanel(): Chainable<Element>
+     checkMapIcon(): Chainable<Element>
+     mockGeolocation(): Chainable<Element>
```
Comment (Proantagonist) [inline apps/aggiemap-angular-e2e/src/support/index.d.ts:27]: All of these command typings shouldn't be in this file. Move back to commands. 

Also, document your functions and make sure the type returns are accurate.

**Review by Proantagonist — CHANGES_REQUESTED** [review]
# Efficiency

The name of the game is meaningful, deterministic, and fast testing. Kill unnecessary explicit `cy.wait()` commands and remove command timeout overrides. Often times there's very little reason to add a timeout as Cypress does it's magic and has ample default timeouts for certain commands that they've come decided on from their own research on human interaction and behavior. If a timeout is "needed" somewhere, there's a chance there's a deeper problem that can be solved by refactoring the selectors, assertions, or stubs. Always aim for having the lowest test runtime.

# Refactor

Go back through the existing tests that you didn't originally author and remove those that don't test anything meaningful with respect to the page/component being tested.

Follow the three A's:

- Arrange: Setup the application state for test suites. The exact same conditions you want for all the test suites (beforeEach).
- Act: Perform an action (click a button, navigate a page, type something).
- Assert: Test expected outcome.

Following the three A's doesn't mean you should do only do a single assertion per test; that's unit testing. We are not unit testing. We are performing varying degree of integration testing. It's OK to have multiple assertions in a test but find the balance between not testing enough and testing too much in a test.

# Spec structuring

Separate and rename spec files to reflect what they are testing. Names like the following are not specific enough:

- feature-data.spec.ts
- routing-data.spec.ts

If needed, add folders and organize by page/category/components. Also ensure that suites in a spec file aren't stepping over the toes of another spec suite's (search component testing inputs -> selection -> popup content -> clicking on directions -> triggering page change -> selecting second trip endpoint -> changing routing modes). Ensure that the scope of the spec stays within its boundaries. Flow tests ARE valid but not when they start becoming overly specific in unnecessary details or the same test is being performed in other spec suite's with varying degrees of details. A scoped spec should test a specific part of an application and do it well.

# Avoid building state through the UI

In the example above, there were many steps required to setup UI state to eventually make an assertion to test an outcome. Each step has the potential to mutate the DOM and internal application state in ways that are not conducive to deterministic tests. Initial state should be set programmatically as much as possible and that same state should be reused for many tests. Flow tests are important but they are not thorough enough nor practical enough to write to be useful to perform controlled tests on all the individual smaller parts that are glossed over to validate a *certain and specific* flow.

# Use "should" language descriptions/context that flows with `it` action descriptions

A `describe` description represents the suite being tested. A suite is a unit and a unit can be as small or large as needed. For example, a unit be a page (landing, login, registration, user details), a flow (login login/out, trip routing creation), a component (sidebar, layer list, legend). The nested `it` title is the test case; something that should be true or pass, or be false or fail. Combined they should aptly describe the test being executed. For example:

- Layer list
  - displays layer list items
  - adds layer to map when clicked
  - children load the correct resource (layer resources)
- Sidebar
  - displays tabs
  - toggles views when a tab is clicked

# Tests that test the Truthy, Falsy, and In-Between

Tests that test the expected are only half of an application. We hope the users see this side more often times than not, but the unfortunate reality is that shit happens and the UI has to be able to react to failed or pending states. What's a failed state? A search input that yielded no results or search that has taken unreasonably long. A popup with missing sections due to lack of information? An entry url that failed to load or failed silently because of invalid inputs? Pending routing results, failed routing results. All of these cases *should* provide some kind of user feedback and just as important as it is to ensure that users see the successful states of the application, we have to make sure they also see the failed and pending states. If there are no such states in place, the discovery of their absence is also tests doing their job.

# Code State at Time of a PR

When you create a PR, you're declaring that your changes for the scope of the PR to be final, working, and polished as best as reasonably possible. Ensure your changes adhere to the [style guide](https://github.com/TamuGeoInnovation/Tamu.GeoInnovation.js.monorepo/wiki/Style-Guide). Remove dead code that was perhaps used for testing variations of a test and don't leave it behind with the excuse that maybe it's be needed in the future or as a just-in-case. That's what version control is for. If it's not being used, it doesn't need to be there.

Close up any TODO's if they're within the scope of the PR. Some TODO's are OK, such as for example trying to build a test but the current implementation doesn't behave as expected. The changes to the tested system would likely be outside the scope of the PR so in that case a TODO to complete test when the system is fixed is acceptable.

Prettier. Format all your files with prettier (covered in the style guide). White-space changes due to formatting are not functional changes and make pull requests a pain in the ass because it adds a lot of noise in the diffs.

Here is some *solid* writing I found: https://ruleoftech.com/2019/notes-of-best-practices-for-writing-cypress-tests

File: apps/aggiemap-angular-e2e/src/fixtures/walk-route-data copy.json [hunk 4]
```diff
@@ -0,0 +1,185 @@
+{
```
[the remark this answers is not in the document]
Comment (Minhb171) [reply apps/aggiemap-angular-e2e/src/fixtures/walk-route-data copy.json]: File removed. Please let me know if this change is not shown.

File: apps/aggiemap-angular-e2e/src/integration/features-data.spec.ts [hunk 5]
```diff
@@ -1,6 +1,8 @@
 /// <reference path="../support/index.d.ts" />
+import * as Zoom from "esri/widgets/Zoom";
```
[the remark this answers is not in the document]
Comment (Minhb171) [reply apps/aggiemap-angular-e2e/src/integration/features-data.spec.ts:2]: Unused imports have been removed.

File: apps/aggiemap-angular-e2e/src/integration/resolutions.ts [hunk 6]
```diff
@@ -1,2 +1,2 @@
 export const desktopSizes = [[1920, 1080], [1366, 768], [1440, 900]];
-export const mobileSizes = [[375, 667], [375, 812], [414, 896], [414, 846], [360, 760], [768, 1024], [768, 1024], [320, 480]]
\ No newline at end of file
+export const mobileSizes = [[375, 812], [768, 1024], [411, 731], [360, 640]]
```
[the remark this answers is not in the document]
Comment (Minhb171) [reply apps/aggiemap-angular-e2e/src/integration/resolutions.ts]: I have removed other resolution list to only test for one resolution for both desktop and mobile resolution.

**Pull request closed without merging**